### PR TITLE
[SPARK-28040][R] Add serialization for glue type

### PR DESCRIPTION
**Pull request by MichaelChirico**
## What changes were proposed in this pull request?

`glue`-class string can be properly used in `sql` (it's really just a `character`)

## How was this patch tested?

`glue` package creates decorated `character` objects:

```
# from ?glue::glue
name <- "Fred"
age <- 50
anniversary <- as.Date("1991-10-12")
glue_obj = glue('My name is {name},',
  'my age next year is {age + 1},',
  'my anniversary is {format(anniversary, "%A, %B %d, %Y")}.')
dput(glue_obj)
# structure("My name is Fred,my age next year is 51,my anniversary is Saturday, October 12, 1991.", class = c("glue", "character"))
```

i.e., it's just `character` with an extra `class`.

`writeObject` and `getSerdeType` both rely on the first class of the input: `type <- class(object)[[1L]]`. Hence for `glue` objects, `type` will be `'glue'`.

Then the `switch` statement will e.g. work like:

```
switch('glue',
         NULL = writeVoid(con),
         integer = writeInt(con, object),
         glue = , # nolint
         character = 'matched to character')
# [1] "matched to character"
```

WIP PR at #24888 will provide a much more general fix to the `glue`-specific problem solved in this PR, but as noted in the comments, `glue` is a pretty common package so this is likely to serve most related issues with a much simpler fix.

### CODE REVIEW

File: R/pkg/R/serialize.R [hunk 1]
```diff
@@ -71,18 +69,19 @@ writeObject <- function(con, object, writeType = TRUE) {
   switch(serdeType,
          NULL = writeVoid(con),
          integer = writeInt(con, object),
+         glue =,
          character = writeString(con, object),
          logical = writeBoolean(con, object),
-         double = writeDouble(con, object),
+         double =,
```
Comment (MichaelChirico) [inline R/pkg/R/serialize.R]: R does cascading `switch` (like C), so e.g. `'double'` will "fall down the ladder" til the next non-empty entry, namely, `numeric`

**Comment by MichaelChirico** [comment]
@HyukjinKwon re: https://issues.apache.org/jira/browse/SPARK-28040

TBH the whole `serialize.R` logic could go for a revamp to be using S3 or S4 generics instead of the current logic which requires relying on the first element of an object's `class`. e.g. `writeType` could get

```
writeType = function(object, con) UseMethod('writeType')
writeType.default = function(object, con) stop('Unsupported type for serialization', class(object))

writeType.integer = function(object, con) writeBin(as.raw(0x69), con) # charToRaw('i')
...
```

Didn't have time for that at the moment so went with the hackish solution here. If you're amenable I can try to spend the time for the more general approach outlined here.

File: R/pkg/R/serialize.R [hunk 2]
```diff
@@ -71,18 +69,19 @@ writeObject <- function(con, object, writeType = TRUE) {
   switch(serdeType,
          NULL = writeVoid(con),
          integer = writeInt(con, object),
+         glue =,
```
Comment (felixcheung) [inline R/pkg/R/serialize.R]: can this be done without hardcoding "glue" since it's not an explicit dependency or a R base type?

Comment (MichaelChirico) [reply R/pkg/R/serialize.R]: That's what would be accomplished by moving this logic to S3 or S4 -- `glue` would end up being handled by `writeType.character` (etc)

Comment (felixcheung) [reply R/pkg/R/serialize.R]: ok that's fair. I'm ok with what you have now. could you open a jira about s3 writeType?

**Comment by felixcheung** [comment]
Jenkins, ok to test

File: R/pkg/R/serialize.R (hunk 2 above)
Comment (MichaelChirico) [reply R/pkg/R/serialize.R]: Yes. I'm about halfway done working on the switch to S3 also.

Comment (MichaelChirico) [reply R/pkg/R/serialize.R]: Filed: https://issues.apache.org/jira/browse/SPARK-28070

File: R/pkg/R/serialize.R (hunk 1 above)
Comment (HyukjinKwon) [inline R/pkg/R/serialize.R]: I think it's better to keep it anyway. Looks previous one is better to read.

File: R/pkg/R/serialize.R [hunk 3]
```diff
@@ -36,15 +36,15 @@
 # nolint end
 
 getSerdeType <- function(object) {
-  type <- class(object)[[1]]
-  if (is.atomic(object) & !is.raw(object) & length(object) > 1) {
+  type <- class(object)[[1L]]
```
Comment (HyukjinKwon) [inline R/pkg/R/serialize.R]: btw why did we change 1 -> 1L? Is it related with the fix?

Comment (MichaelChirico) [reply R/pkg/R/serialize.R]: no, just a microoptimization while i'm in this source

File: R/pkg/R/serialize.R (hunk 1 above)
Comment (MichaelChirico) [reply R/pkg/R/serialize.R]: readability is in the eye of the beholder 🙃 

anyway this `switch` will go away after 28070

**Comment by MichaelChirico** [comment]
Have also filed https://github.com/apache/spark/pull/24888; leaving this PR open for now since the latter is a lot more involved / may take a tad longer to work out all the kinks

File: R/pkg/R/serialize.R (hunk 1 above)
Comment (HyukjinKwon) [reply R/pkg/R/serialize.R]: k. but let's keep only related changes in case we backport or revert. that will likely reduce the conflicts when we do.

**Comment by HyukjinKwon** [comment]
ok to test

**Comment by MichaelChirico** [comment]
Have added a commit to address the complaints of the linter. Should be ready to merge.

**Comment by MichaelChirico** [comment]
OK... a bit of a Catch-22 for the linter:

```
R/serialize.R:72:17: style: Commas should never have a space before.
         glue = ,
               ~^
# etc
```

But also from previous build:

```
R/serialize.R:72:15: style: Put spaces around all infix operators.
         glue =,
              ^~
```

Shall I revert to explicit behavior of `switch`? Or fine to ignore...?

File: R/pkg/R/serialize.R [hunk 4]
```diff
@@ -71,18 +69,19 @@ writeObject <- function(con, object, writeType = TRUE) {
   switch(serdeType,
          NULL = writeVoid(con),
          integer = writeInt(con, object),
+         glue = ,
```
Comment (HyukjinKwon) [inline R/pkg/R/serialize.R]: Can we just explicitly write `writeString(con, object)`?

Comment (MichaelChirico) [reply R/pkg/R/serialize.R]: Yes that's what I meant by `explicit behavior`... Just doesn't feel very "R" to me. At a quick glance I don't see anything mentioned in the tidyverse or Google R style guides.

Comment (MichaelChirico) [reply R/pkg/R/serialize.R]: I ran a quick benchmark... seems to me using implicit `switch` is about 40% faster:

```
library(microbenchmark)
set.seed(38459)
options = sprintf(
  "'%s'",
  replicate(1e4, paste(sample(letters, sample(5L, 1L), TRUE), collapse = ''))
)

nrep = 1e7
timings = matrix(nrow = nrep, ncol = 3L)

for (ii in 1:nrep) {
  nopt = sample(20L, 1L)
  thisopt = sample(options, nopt)
  thisval = paste0(1:nopt)
  blank = sample(nopt, sample(nopt, 1L))
  headval = sample(options, 1L)
  
  fullexpr = parse(
    text = sprintf("switch(%s, %s)", headval,
                   paste(sprintf('%s=%s', thisopt, thisval), collapse = ', '))
  )
  blankexpr = parse(
    text = sprintf("switch(%s, %s)", headval,
                   paste(sprintf('%s=%s', thisopt, `[<-`(thisval, blank, '')),
                         collapse = ', '))
  )
  
  t0 = get_nanotime()
  eval(fullexpr)
  t1 = get_nanotime()
  eval(blankexpr)
  t2 = get_nanotime()
  
  timings[ii, 1L] = nopt
  timings[ii, 2L] = t1 - t0
  timings[ii, 3L] = t2 - t1
}
```

with

```
summary(timings[ , 3L]/timings[ , 2L])
     Min.   1st Qu.    Median      Mean   3rd Qu.      Max.      NA's 
     0.00      0.57      0.60      0.77      0.64 137407.00    227997
```

(I cut the benchmark short by 300K repetitions)

Comment (HyukjinKwon) [reply R/pkg/R/serialize.R]: Hm, interesting.

I believe you can skip linter as:
```R
statement  # nolint
```

or 

```R
# nolint start
statement
# nolint end
```

Comment (MichaelChirico) [reply R/pkg/R/serialize.R]: oh ok, let me try

**Comment by MichaelChirico** [comment]
OK finally. @HyukjinKwon ready to merge?

**Comment by HyukjinKwon** [comment]
I don't know enough context about `glue` and I don't know if that library is popular and widely used enough to support in Spark.

If it is difficult to test, it might be better to clarify how you tested it manually in the PR description.

Let me leave it to @felixcheung

**Comment by MichaelChirico** [comment]
I would say it's pretty popular and widely used. It's part of the `tidyverse`:

https://github.com/tidyverse/glue

It has 165 direct reverse-depends (packages importing, depending, or suggesting `glue`):

```
devtools::revdep('glue', c('Depends', 'Imports', 'Suggests'), bioconductor = TRUE)
```

And recursively 13,997 packages depend on `glue` (6,016 of which are directly, i.e. excluding `Suggest`ed dependencies).

Will add a quick note to the PR description.

**Comment by zero323** [comment]
> I don't know enough context about `glue` and I don't know if that library is popular and widely used enough to support in Spark.

`glue` pretty much become a standard way of building strings in R these days, so making SparkR aware of it, at least on the driver side, is a good idea.

**Pull request closed without merging**